### ISSUE TITLE
Shell script being /wildly/ misclassified as C++

### ISSUE THREAD

**Issue by Orc** [github/linguist#3569]
Part of my autoconfiguration code is being classified as C++;  <https://github.com/Orc/configure/blob/6509031ef4fa1d2ff267d4650abced0aa76311a5/configure.inc> in particular.    C++ is a pretty bizarre language, but both the c++ compilers I've got access to (clang,g++) blow up in a spectacular fashion when fed configure.inc as input.

**Comment by lildude** [comment]
Looking at the debug output when run locally:

```
$ LINGUIST_DEBUG=1 bin/linguist ~/tmp/trash/configure/configure.inc
configure.inc: 1935 lines (1631 sloc)
  type:      Text
  mime type: text/plain
  Assembly = -21005.306 +  -5.682 = -21010.989
       C++ = -19802.735 +  -3.668 = -19806.402
      HTML = -22537.331 +  -5.682 = -22543.014
      PAWN = -22657.060 +  -6.781 = -22663.841
       PHP = -20589.898 +  -4.530 = -20594.428
POV-Ray SDL = -24026.176 +  -4.989 = -24031.165
    Pascal = -21300.928 +  -5.172 = -21306.099
       SQL = -23512.827 +  -4.989 = -23517.817
SourcePawn = -20199.204 +  -5.682 = -20204.887
  language:  C++
$
```

... I can see that Linguist has made it all the way through to the classifier when attempting to determine the language used here and this has determined the language to be C++.  The classifier is the "last resort, lets take a guess at it" step and is known to be far from perfect.

It didn't even take a stab at guessing Shell as the `.inc` extension isn't currently associated with the Shell language. It is however associated with all those shown above.  Adding the `.inc` extension _would_ result in the correct language, however I don't believe it is the correct thing to do here as AFAIK `.inc` in not an extension commonly associated with shell scripts.

Accordingly, a better solution to get you correct language detection is to use a [manual override](https://github.com/github/linguist#overrides), or change the extension to one commonly used with shell scripts, or add a shebang to the top of the script.

**Comment by Orc** [comment]
Yes, it's not being classified properly; the problem is that linguist's failure case of "look at the file extension" is pretty much exactly the wrong choice.  And not just in this case; in the general case of trying to classify a file the fallback should just be "text", not something that it's already ruled out.

**Comment by Orc** [comment]
I think that we're talking past each other.  What I'm saying is that the default behavior of throwing up its hands in the air and randomly picking a language from a handful of wildly improbable choices is ridiculous and a better behavior would be to pick a floor where if your language guesser can't clear that level of certainty it should just say the offending file is unknown.

**Comment by pchaigno** [comment]
> pick a floor where if your language guesser can't clear that level of certainty it should just say the offending file is unknown.

Definitely! We've talked about this several times in the past and it would be a first, necessary step towards an *unknown* option. This would require careful changes to the naive Bayesian classifier for which none of us has much time currently, but if you want to contribute something in that direction, we'd welcome it!
